src/devices/src/virtio/vsock/event_handler.rs

```diff
@@ -113,70 +113,30 @@ where
-        let self_subscriber = event_manager
-            .subscriber(self.activate_evt.as_raw_fd())
-            .unwrap();
+        let self_subscriber = match event_manager.subscriber(activate_fd) {
+            Ok(subscriber) => subscriber,
+            Err(e) => {
+                error!("Failed to process vsock activate evt: {:?}", e);
+                return;
+            }
+        };
+
+        // Interest list changes when the device is activated.
+        let interest_list = self.interest_list();
+        for event in interest_list {
+            event_manager
+                .register(event.data() as i32, event, self_subscriber.clone())
+                .unwrap_or_else(|e| {
+                    error!("Failed to register vsock events: {:?}", e);
+                });
+        }
 
-        event_manager
-            .register(
-                self.queue_events[RXQ_INDEX].as_raw_fd(),
-                EpollEvent::new(
-                    EventSet::IN,
-                    self.queue_events[RXQ_INDEX].as_raw_fd() as u64,
-                ),
-                self_subscriber.clone(),
-            )
-            .unwrap_or_else(|e| {
-                error!("Failed to register vsock rxq with event manager: {:?}", e);
-            });
-
-        event_manager
-            .register(
-                self.queue_events[TXQ_INDEX].as_raw_fd(),
-                EpollEvent::new(
-                    EventSet::IN,
-                    self.queue_events[TXQ_INDEX].as_raw_fd() as u64,
-                ),
-                self_subscriber.clone(),
-            )
-            .unwrap_or_else(|e| {
-                error!("Failed to register vsock txq with event manager: {:?}", e);
-            });
-
-        event_manager
-            .register(
-                self.queue_events[EVQ_INDEX].as_raw_fd(),
-                EpollEvent::new(
-                    EventSet::IN,
-                    self.queue_events[EVQ_INDEX].as_raw_fd() as u64,
-                ),
-                self_subscriber.clone(),
-            )
-            .unwrap_or_else(|e| {
-                error!("Failed to register vsock evq with event manager: {:?}", e);
-            });
-
-        event_manager
-            .register(
-                self.backend.as_raw_fd(),
-                EpollEvent::new(
-                    self.backend.get_polled_evset(),
-                    self.backend.as_raw_fd() as u64,
-                ),
-                self_subscriber,
-            )
-            .unwrap_or_else(|e| {
-                error!("Failed to register vsock backend events: {:?}", e);
-            });
-
-        event_manager
-            .unregister(self.activate_evt.as_raw_fd())
-            .unwrap_or_else(|e| {
-                error!("Failed to unregister vsock activate evt: {:?}", e);
-            })
+        event_manager.unregister(activate_fd).unwrap_or_else(|e| {
+            error!("Failed to unregister vsock activate evt: {:?}", e);
+        });
```

Comment:
`handle_activate_event` should be `process_activate_event` to match the corresponding methods in the other devices. Or even better, we could define an interface for this since the implementation is exactly the same for all devices.

Reply:
An interface would definitely be nice for code de-duplication, but the snag I hit when trying to do it is that it doesn't really belong in either `VirtioDevice` or `Subscriber`; it would have to be a `trait SubscriberVirtioDevice: Subscriber + VirtioDevice` and it would live in a separate file under devices so that we don't make a mess of things so the end result adds enough extra complexity that I just feel it's not worth it.

I'm not against it, but I don't see a nice, clear and simple way to do it.

Reply:
I think this is fine and more elegant than having the same method in 3 places (or more if we're going to add devices). We can log an issue for this improvement in order to think a bit more about it.